src/envoy/http/authn/http_filter_test.cc

```diff
@@ -0,0 +1,144 @@
+AuthenticatorBase* createAlwaysPassAuthenticator(
+    FilterContext* filter_context,
+    const AuthenticatorBase::DoneCallback& done_callback) {
+  class _local : public AuthenticatorBase {
+   public:
+    _local(FilterContext* filter_context,
+           const AuthenticatorBase::DoneCallback& callback)
+        : AuthenticatorBase(filter_context, callback) {}
+    void run() override { done(true); }
+  };
+  return new _local(filter_context, done_callback);
+}
+
+class MockAuthenticationFilter : public AuthenticationFilter {
+ public:
+  // We'll use fake authenticator for test, so policy is not really needed. Use
+  // default policy for simplicity.
+  MockAuthenticationFilter()
+      : AuthenticationFilter(
+            istio::authentication::v1alpha1::Policy::default_instance()) {}
+  ~MockAuthenticationFilter(){};
+
+  MOCK_METHOD2(createPeerAuthenticator,
+               AuthenticatorBase*(FilterContext*,
+                                  const AuthenticatorBase::DoneCallback&));
+  MOCK_METHOD2(createOriginAuthenticator,
+               AuthenticatorBase*(FilterContext*,
+                                  const AuthenticatorBase::DoneCallback&));
+};
+
+class AuthentiationFilterTest : public testing::Test {
+ public:
+  AuthentiationFilterTest()
+      : request_headers_{{":method", "GET"}, {":path", "/"}} {}
+  ~AuthentiationFilterTest() {}
+
+  void SetUp() override {
+    filter_.setDecoderFilterCallbacks(decoder_callbacks_);
+  }
+
+ protected:
+  StrictMock<MockAuthenticationFilter> filter_;
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> decoder_callbacks_;
+  Http::TestHeaderMapImpl request_headers_;
+};
+
+TEST_F(AuthentiationFilterTest, PeerFail) {
+  // Peer authentication fail, request should be rejected with 401. No origin
+  // authentiation needed.
+  EXPECT_CALL(filter_, createPeerAuthenticator(_, _))
+      .Times(1)
+      .WillOnce(Invoke(createAlwaysFailAuthenticator));
+  EXPECT_CALL(decoder_callbacks_, encodeHeaders_(_, _))
+      .Times(1)
+      .WillOnce(testing::Invoke([](Http::HeaderMap& headers, bool) {
+        EXPECT_STREQ("401", headers.Status()->value().c_str());
+      }));
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_.decodeHeaders(request_headers_, true));
+}
+
+TEST_F(AuthentiationFilterTest, PeerPassOrginFail) {
+  // Peer pass thus origin authentication must be called. Final result should
+  // fail as origin authn fails.
+  EXPECT_CALL(filter_, createPeerAuthenticator(_, _))
+      .Times(1)
+      .WillOnce(Invoke(createAlwaysPassAuthenticator));
+  EXPECT_CALL(filter_, createOriginAuthenticator(_, _))
+      .Times(1)
+      .WillOnce(Invoke(createAlwaysFailAuthenticator));
+  EXPECT_CALL(decoder_callbacks_, encodeHeaders_(_, _))
+      .Times(1)
+      .WillOnce(testing::Invoke([](Http::HeaderMap& headers, bool) {
+        EXPECT_STREQ("401", headers.Status()->value().c_str());
+      }));
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
```

Comment:
please test request has been suspended case, (by making remote call)

Reply:
Doesn't expect decoder_callbacks call encodeHeader (send 401) above do the job?